community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/transitiveClosure.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.cypher.internal.frontend.v3_4.ast.rewriters
+
+import org.neo4j.cypher.internal.frontend.v3_4.ast.Where
+import org.neo4j.cypher.internal.frontend.v3_4.phases.{BaseContext, Condition}
+import org.neo4j.cypher.internal.util.v3_4.Foldable._
+import org.neo4j.cypher.internal.util.v3_4.{Rewriter, bottomUp, topDown}
+import org.neo4j.cypher.internal.v3_4.expressions._
+
+import scala.annotation.tailrec
+
+/**
+  * Transitive closure of where clauses.
+  *
+  * Given a where clause, `WHERE a.prop = b.prop AND b.prop = 42` we rewrite the query
+  * into `WHERE a.prop = 42 AND b.prop = 42`
+  */
+case object transitiveClosure extends StatementRewriter {
+
+  override def description: String = "transitive closure in where clauses"
+
+  override def instance(ignored: BaseContext): Rewriter = transitiveClosureRewriter
+
+  override def postConditions: Set[Condition] = Set.empty
+
+  private case object transitiveClosureRewriter extends Rewriter {
+
+    def apply(that: AnyRef): AnyRef = instance.apply(that)
+
+    private val instance: Rewriter = topDown(Rewriter.lift {
```

Comment:
Is there a reason we do this `topDown` instead of `bottomUp`? I was under the impression that `bottomUp` is preferable from a performance perspective, if either work.